dev/com.ibm.ws.security.wim.adapter.ldap_fat.krb5/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/krb5/ApacheDSandKDC.java

```diff
@@ -0,0 +1,649 @@
+        assertNotNull("LdapServer is null and cannot be started", ldapServer);
+        assertNotNull("DirectoryService is null and cannot be started", directoryService);
+
+        Log.info(c, "startAllServers", "Starting all ApacheDS servers");
+        directoryService.startup();
+        ldapServer.start();
+        kdcServer.start();
+        Log.info(c, "startAllServers", "Started all ApacheDS servers");
+    }
+
+    /**
+     * Get a free port.
+     *
+     * @return The free port.
+     * @throws IOException If a free port could not be found.
+     */
+    private static int getOpenPort() throws IOException {
+        Log.info(c, "getOpenPort", "Entry.");
+        ServerSocket s = new ServerSocket(0);
+        Log.info(c, "getOpenPort", "Got socket.");
+        int port = s.getLocalPort();
+        Log.info(c, "getOpenPort", "Got port " + s);
+        s.close();
+        Log.info(c, "getOpenPort", "Close socket");
+        return port;
+    }
+
+    public static DirectoryService getDirectoryService() {
+        return directoryService;
+    }
+
+    /**
+     * Create the default krb config file.
+     *
+     * @return
+     * @throws IOException
+     */
+    public static String createConfigFile() throws IOException {
+        Log.info(c, "createConfigFile", "Creating config file");
+        File ccFile = File.createTempFile(bindUserName + "krb5-", ".conf");
+        if (!FAT_TEST_LOCALRUN) {
+            ccFile.deleteOnExit();
+        }
+
+        Path outputPath = Paths.get(ccFile.getAbsolutePath());
+        Log.info(c, "generateConf", "creating krb.conf file.");
+        String conf = "[libdefaults]\n" +
+                      "        rdns = false\n" +
+                      "        dns_lookup_realm = false\n" +
+                      "        udp_preference_limit = 1\n" +
+                      "        dns_lookup_kdc = false\n" +
+                      "        renew_lifetime = 7d\n" +
+                      "        forwardable = true\n" +
+                      "        default_realm = " + DOMAIN.toUpperCase() + "\n" +
+                      "\n" +
+                      "[realms]\n" +
+                      "        " + DOMAIN.toUpperCase() + " = {\n" +
+                      "                kdc = " + ldapServerHostName + ":" + KDC_PORT + "\n" +
+                      "        }\n" +
+                      "\n" +
+                      "[domain_realm]\n" +
+                      "        ." + DOMAIN.toLowerCase() + " = " + DOMAIN.toUpperCase() + "\n" +
+                      "        " + DOMAIN.toLowerCase() + " = " + DOMAIN.toUpperCase() + "\n";
+        outputPath.getParent().toFile().mkdirs();
+        Files.write(outputPath, conf.getBytes(StandardCharsets.UTF_8));
+
+        Log.info(c, "createConfigFile", "krb.conf file contents: " + conf);
+        return ccFile.getAbsolutePath();
+    }
+
+    public static String createConfigFileWithBadKDCPort() throws IOException {
```

Comment:
Probably can combine these two methods that create the config file. Maybe have a createKrb5ConfigFile(Integer port) method that will set the port if it is specified, otherwise it will use KDC_PORT (or you can always pass in the port. Will save a bunch of extra code here I think.

Reply:
Good point -- I think I planned to have some other variations and then didn't come back to it.